Move info on KOTS min and target version options (#252)
replicatedhq/replicated-docs: docs/vendor/packaging-kots-versions.md

# Setting Minimum and Target Versions for KOTS

> Minimum and target KOTS versions are optional features that are configured in the Application custom resource file. The attributes in this file help to control the application experience but are not deployed to the cluster. For more information, see [Application Custom Resource](../reference/custom-resource-application).

Author: It was copy/pasted from the first paragraph on this page, to give context as to how the Application custom resource file works as a whole. https://docs.replicated.com/vendor/packaging-custom-resources

> Minimum and target KOTS versions are optional features that are configured in the Application custom resource file. The attributes in this file help to control the application experience but are not deployed to the cluster. For more information, see [Application Custom Resource](../reference/custom-resource-application).

Author: I could clarify it be saying "The attributes in the Application custom resource help to control...." How does that sound to you?

> Minimum and target KOTS versions are optional features that are configured in the Application custom resource file. The attributes in this file help to control the application experience but are not deployed to the cluster. For more information, see [Application Custom Resource](../reference/custom-resource-application).

Reviewer: That link was helpful, thanks. The custom resources themselves aren't deployed to the cluster, whereas this makes it sound like the attributes aren't deployed to the cluster. I think this might be clearer.
```suggestion
Minimum and target KOTS versions are optional features that are configured in the Application custom resource file. The attributes in this custom resource help to control the application experience, but the custom resource is not deployed to the cluster. For more information, see [Application Custom Resource](../reference/custom-resource-application).
```

:::note

> The Replicated app manager is based on the KOTS open source project. The KOTS version is the same as the app manager version. For example, KOTS v1.60 is the same as the app manager v1.60. For more information, see [KOTS add-on](https://kurl.sh/docs/add-ons/kotsadm) in the open source kURL documentation.

Reviewer: I'll put my suggestion here again, but let me know if I'm missing a reason to keep this here.
```suggestion
The Replicated app manager is based on the KOTS open source project. The KOTS version is the same as the app manager version. For example, KOTS v1.60 is the same as the app manager v1.60.
```

:::

## Using Minimum KOTS Versions (Beta)

>Introduced in app manager v1.62.0.

> The `minKotsVersion` attribute in the Application manifest file defines the minimum KOTS version that is required by the application release. This can be useful when you want to get users who are lagging behind to update to a more recent KOTS version, or if your application requires functionality that was introduced in a particular KOTS version.

Reviewer: This feels like more direct language. Let me know if you agree.
```suggestion
The `minKotsVersion` attribute in the Application manifest file defines the minimum KOTS version that is required by the application release. This is useful when you want to get users who are lagging behind to update to a more recent KOTS version, or if your application requires functionality that was introduced in a particular KOTS version.
```

Including this attribute enforces compatibility checks for both new installations and application updates. An installation or update is blocked if the currently deployed KOTS version is earlier than the specified minimum KOTS version. Users must upgrade to at least the specified minimum version of KOTS before they can install or update the application.

> For more information about the user experience for updates, see [How the Admin Console Handles `minKotsVersion`](#how-the-admin-console-handles-minkotsversion).

Reviewer: ```suggestion
For more information about the user experience when a minimum KOTS version is specified, see [How the Admin Console Handles `minKotsVersion`](#how-the-admin-console-handles-minkotsversion).
```

Reviewer: Also just curious if this is necessary since this section is right beneath it and we don't have a similar note for target KOTS version

> For more information about the user experience for updates, see [How the Admin Console Handles minKotsVersion](#how-the-admin-console-handles-minkotsversion).

Author: We do but I don't think they're supported in MD titles/headings bc it was migrated without backticks. Not sure if they're supported in hyperlinking either, but I can give it a try.

### How the Admin Console Handles minKotsVersion

> When you promote a new release specifying a minimum KOTS version that is later than what a user currently has deployed, that application version appears in the version history of the admin console after the user checks for updates. However, it is not downloaded.

Reviewer: I got confused for a second when reading this, so I feel it's clearer to mention the check for updates sooner.
```suggestion
When you promote a new release specifying a minimum KOTS version that is later than what a user currently has deployed, and that user checks for updates, that application version appears in the version history of the admin console. However, it is not downloaded.
```

> The admin console temporarily displays an error message that informs the user that they must update KOTS before downloading the application version. This error also displays when the user checks for updates with the [`kots upstream upgrade`](../reference/kots-cli-upstream-upgrade) command. KOTS cannot update itself automatically, and users cannot update KOTS from the admin console.

Reviewer: Now that we have better docs on updating KOTS/app manager, we could link to those here.

Also, would this now be too long as one paragraph? I split it up just in case, but that's up to you.
```suggestion
The admin console temporarily displays an error message that informs the user that they must update KOTS before downloading the application version. This error also displays when the user checks for updates with the [`kots upstream upgrade`](../reference/kots-cli-upstream-upgrade) command.

KOTS cannot update itself automatically, and users cannot update KOTS from the admin console. For more information on updating KOTS, see [Updating the Admin Console on an Existing Cluster](../enterprise/updating-existing-cluster) and [Updating the Admin Console on a Kubernetes Installer-Created Cluster](../enterprise/updating-embedded-cluster).
```

> After updating KOTS to the minimum version or later, users can use the admin console or use the [`kots upstream download`](../reference/kots-cli-upstream-download) command to download the release.

Reviewer: ```suggestion
After updating KOTS to the minimum version or later, users can use the admin console or use the [`kots upstream download`](../reference/kots-cli-upstream-download) command to download the release and subsequently deploy it.
```

> After updating KOTS to the minimum version or later, users can use the admin console or use the [`kots upstream download`](../reference/kots-cli-upstream-download) command to download the release and subsequently deploy it.

Reviewer: ```suggestion
After updating KOTS to the minimum version or later, users can use the admin console or the [`kots upstream download`](../reference/kots-cli-upstream-download) command to download the release and subsequently deploy it.
```

## Using Target KOTS Versions

>Introduced in app manager v1.62.0.

>  Including `targetKotsVersion` in the Application manifest file enforces compatibility checks for new installations. It blocks the installation if a user tries to install a version that is later than the target version. For example, this can prevent users from installing a version of KOTS that you have not tested yet.

Reviewer: ```suggestion
 Including `targetKotsVersion` in the Application manifest file enforces compatibility checks for new installations. It blocks the installation if a user tries to install a version of KOTS that is later than the target version. For example, this can prevent users from installing a version of KOTS that you have not tested yet.
```

If the latest release in a channel includes `targetKotsVersion`, the install command for existing clusters is modified to install that specific version of KOTS. The install command for existing clusters is on the channel card in the [vendor portal](https://vendor.replicated.com).

> ### How the Admin Console Handles targetKotsVersion

Author: The CSS handles this in an ugly way, so we're not adding the backticks in titles/headings right now. Added it to a story in Shortcut since there are other CSS changes that we'd like to have.

Specifying a `targetKotsVersion` does not prevent an end user from upgrading to a later version of KOTS after the initial installation.

If a new version of the application specifies a later target KOTS version than what is currently installed, users are not prevented from deploying that version of the application.

> If a user's admin console is running a version of KOTS that is earlier than the target version specified in the new version of the application, the admin console displays a banner indicating that a newer supported version of KOTS is available.

Reviewer: ```suggestion
If a user's admin console is running a version of KOTS that is earlier than the target version specified in a new version of the application, the admin console displays a notification in the footer, indicating that a newer supported version of KOTS is available.
```

### Using Target Versions with the Kubernetes Installer

For installations to a cluster created by the Replicated Kubernetes installer, the version of the KOTS add-on must not be later than the target KOTS version specified in the Application manifest file. If the KOTS add-on version is later than the version specified for `targetKotsVersion`, the initial installation fails.

> For more information about the KOTS add-on, see [KOTS add-on](https://kurl.sh/docs/add-ons/kotsadm) in the open source kURL documentation.

Reviewer: This is a good place to mention this since we're specifically talking about kURL clusters.

Author: Can you clarify where you think that needs to go? For example, it does say "For installations to a cluster created by the Replicated Kubernetes installer"...but I could make the two paragraphs into one, which might make that clearer. Would that help?

Reviewer: I wasn't requesting any change, just affirming that this is a good example of mentioning the KOTS add-on, since this is in fact about kURL clusters.

Author: Thanks!
